paasta_tools/frameworks/task_store.py

```diff
@@ -0,0 +1,180 @@
+from __future__ import unicode_literals
+
+import json
+
+from kazoo.client import KazooClient
+from kazoo.exceptions import NoNodeError
+
+from paasta_tools.utils import _log
+
+
+class MesosTaskParametersIsImmutableError(Exception):
+    pass
+
+
+class MesosTaskParameters(object):
+    def __init__(
+        self,
+        health=None,
+        mesos_task_state=None,
+        is_draining=None,
+        is_healthy=None,
+        offer=None,
+        resources=None,
+    ):
+        self.__dict__['health'] = health
+        self.__dict__['mesos_task_state'] = mesos_task_state
+        self.__dict__['is_draining'] = is_draining
+        self.__dict__['is_healthy'] = is_healthy
+        self.__dict__['offer'] = offer
+        self.__dict__['resources'] = resources
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__
+
+    def __repr__(self):
+        return "%s(\n    %s)" % (type(self).__name__, ',\n    '.join(["%s=%r" % kv for kv in self.__dict__.items()]))
+
+    def __setattr__(self, name, value):
+        raise MesosTaskParametersIsImmutableError()
+
+    def __delattr__(self, name):
+        raise MesosTaskParametersIsImmutableError()
+
+    def merge(self, **kwargs):
+        """Return a merged MesosTaskParameters object, where attributes in other take precedence over self."""
+
+        new_dict = self.__dict__
+        new_dict.update(kwargs)
+
+        return MesosTaskParameters(**new_dict)
+
+    @classmethod
+    def deserialize(cls, serialized_params):
+        return cls(**json.loads(serialized_params))
+
+    def serialize(self):
+        return json.dumps(self.__dict__).encode('utf-8')
+
+
+class TaskStore(object):
+    def __init__(self, service_name, instance_name, framework_id, system_paasta_config):
+        self.service_name = service_name
+        self.instance_name = instance_name
+        self.framework_id = framework_id
+        self.system_paasta_config = system_paasta_config
+
+    def get_task(self, task_id):
+        """Get task data for task_id. If we don't know about task_id, return None"""
+        raise NotImplementedError()
+
+    def get_all_tasks(self):
+        """Returns a dictionary of task_id -> MesosTaskParameters for all known tasks."""
+        raise NotImplementedError()
+
+    def overwrite_task(self, task_id, params):
+        raise NotImplementedError()
+
+    def add_task_if_doesnt_exist(self, task_id, **kwargs):
+        """Add a task if it does not already exist. If it already exists, do nothing."""
+        if self.get_task(task_id) is not None:
+            return
+        else:
+            self.overwrite_task(task_id, MesosTaskParameters(**kwargs))
+
+    def update_task(self, task_id, **kwargs):
+        existing_task = self.get_task(task_id)
+        if existing_task:
+            merged_params = existing_task.merge(**kwargs)
+        else:
+            merged_params = MesosTaskParameters(**kwargs)
+
+        self.overwrite_task(task_id, merged_params)
```

Comment:
I've overridden update_task in ZKTaskStore so that it passes to set() the version received when reading the data, and retries if this fails. I've also better handled the case where someone else created the task between us reading it (and not seeing anything) and trying to create it.